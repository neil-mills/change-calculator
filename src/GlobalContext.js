import React, { createContext, useState } from 'react';
import currencies from './data/currencies';

const initialState = {
  calculateChange: () => {},
  results: null,
};

export const GlobalContext = createContext(initialState);

export const GlobalProvider = ({ children, ...rest }) => {
  const [results, setResults] = useState(initialState.results);

  const calculateChange = ({ currency, price, tendered, json }) => {
    const prefix = {
      gbp: '£',
      usd: '$',
      eur: '€',
    };

    const setPrefix = (value) => `${prefix[currency]}${value}`;

    const totalChange = parseFloat(tendered - price).toFixed(2);
    let balance = Math.round(totalChange * 100);
    const change = currencies
      .filter((c) => c.currency === currency)
      .sort((a, b) => b.denomination - a.denomination)
      .reduce((res, tender) => {
        const { title, denomination } = tender;
        const total = Math.floor(balance / denomination);
        if (total > 0) {
          balance -= total * denomination;
          res = [...res, { title, total }];
        }
        return res;
      }, []);
      const jsonChange = json ? JSON.stringify(change) : null;
    setResults({
      change,
      jsonChange,
      totalChange: setPrefix(totalChange),
      price: setPrefix(price),
      tendered: setPrefix(tendered),
    });
  };

  const values = {
    ...initialState,
    calculateChange,
    results,
    ...rest,
  };

  return (
    <GlobalContext.Provider value={values}>{children}</GlobalContext.Provider>
  );
};
