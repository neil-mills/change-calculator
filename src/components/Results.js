import React, { useContext, Fragment } from 'react';
import { GlobalContext } from '../GlobalContext';

const Results = () => {
  const { results } = useContext(GlobalContext);
  if (!results) return null;
  const { change = [], totalChange = '', price = '', tendered = '', jsonChange = null } = results;
  if (jsonChange) {
    return <p className="title is-4">Coming soon...</p>
  }
  return (
    <section className="box">
      <p className="title is-4">{`You supplied ${tendered}, with a product price of ${price}`}</p>
      {change.length > 0 && (
        <Fragment>
          <p className="title is-5">This is your change:</p>
          <ul data-testid="results" className="title is-5">
            {change.map(({ title, total }, i) => (
              <li key={i} data-testid="result-item">
                <span>{`${total}x ${title}`}</span>
              </li>
            ))}
          </ul>
        </Fragment>
      )}
      <p className="title is-4">Total change: {totalChange}</p>
    </section>
  );
};

export default Results;
