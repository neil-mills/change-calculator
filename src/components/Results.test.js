import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Results from './Results';
import { GlobalProvider } from '../GlobalContext';

const renderComponent = (values = {}) =>
  render(
    <GlobalProvider {...values}>
      <Results />
    </GlobalProvider>
  );

const results = {
  change: [
    { title: '£20', total: 1 },
    { title: '£10', total: 1 },
    { title: '50p', total: 1 },
  ]
};

describe('Results', () => {
  beforeEach(() => {
    cleanup();
  });

  it('does not render if no results in context', () => {
    const mockContext = {
      results: [],
    };
    const { queryByTestId } = renderComponent(mockContext);
    expect(queryByTestId('results')).toBeFalsy();
  });

  it('renders the results', () => {
    const mockContext = {
      results,
    };
    const { asFragment } = renderComponent(mockContext);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders coming soon if json export in results', () => {
    const mockContext = {
      results: {
        ...results,
        jsonChange: []
      }
    }
    const { getByText } = renderComponent(mockContext);
    expect(getByText('Coming soon...')).toBeTruthy();
  })

});
