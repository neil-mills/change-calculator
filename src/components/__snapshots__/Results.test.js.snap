// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders the results 1`] = `
<DocumentFragment>
  <section
    class="box"
  >
    <p
      class="title is-4"
    >
      You supplied , with a product price of 
    </p>
    <p
      class="title is-5"
    >
      This is your change:
    </p>
    <ul
      class="title is-5"
      data-testid="results"
    >
      <li
        data-testid="result-item"
      >
        <span>
          1x £20
        </span>
      </li>
      <li
        data-testid="result-item"
      >
        <span>
          1x £10
        </span>
      </li>
      <li
        data-testid="result-item"
      >
        <span>
          1x 50p
        </span>
      </li>
    </ul>
    <p
      class="title is-4"
    >
      Total change: 
    </p>
  </section>
</DocumentFragment>
`;
